### PR TITLE
Improved accuracy of `IconProps` interface

### PR DESCRIPTION
**Pull request by amcclain**
- Use the `IconName` and `IconPrefix` types provided by FontAwesome.
- Define a `ResolvedIconProps` interface, for specific usages where we know `iconName` and `prefix` have been applied and will be present.

Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile, or determined not required.
- [x] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

**Comment by amcclain** [comment]
I started looking at this due to warnings being thrown around icons that turned out were actually caused by https://github.com/xh/hoist-react/commit/9435122c610de16e770ffb644cce9b83b14f8b3a

But I had started tightening things up and still there is some utility in taking these changes.  My only concern is that the use of the `IconName` exported type from FA might flag false-positive warnings of some kind, or require apps that are passing icon names as vars to more precisely type them.  Not sure that would happen in practice though, or be a blocker - could be helpful...

File: icon/Icon.ts [hunk 1]
```diff
@@ -59,27 +53,50 @@ export interface IconProps extends HoistProps, Partial<Omit<FontAwesomeIconProps
...
+     * your app bootstrap code, eg:
+     * ```
+     *   import {library} from '@fortawesome/fontawesome-svg-core';
+     *   import {faDreamIcon} from '@fortawesome/pro-regular-svg-icons';
+     *   library.add(faDreamIcon);
+     * ```
+     * and then pass its string name to this factory: `icon({iconName: 'dream-icon'})`
      */
-    icon(opts?: IconProps): any {
+    icon(opts: IconProps & {iconName: IconName}): any {
```
Comment (lbwexler) [inline icon/Icon.ts:99]: why is the & {iconName:IconName} needed?

**Review by lbwexler — APPROVED** [review]

**Comment by lbwexler** [comment]
I am for getting this merged -- seems like a nice pickups to get a list of all of the available names. Very easy to cast `as IconName` for any problematic usages.

File: icon/Icon.ts (hunk 1 above)
Comment (lbwexler) [reply icon/Icon.ts:99]: I tried without it, and it seemed to work, but perhaps am missing something?

**Review by ghsolomon — APPROVED** [review]
Looks good to me

**Pull request merged**